### PR TITLE
[ETHEREUM-CONTRACTS] SuperfluidFrameworkDeployer Additions

### PR DESCRIPTION
**Pull request by 0xdavinchee**
## SuperfluidFrameworkDeployer.sol changes
* add deploy native asset super token
* add deploy pure super token
* add tests for token deployments
* modify .prettierrc and tsconfig.json

## SDK-Core Changes
* refactored the test harnessing (use TestEnvironment)

## Subgraph Changes
* **IMPORTANT** For the liquidation test, I delete the existing flow then recreate and run the liquidation test. For some reason, updating then liquidating leads to negative `balanceUntilUpdatedAt` which should never occur.
* new hardcoded addresses (sorry future me or dev who has to work on this)
* fixed up some expect values based on SuperfluidFrameworkDeployer logic (related to wrapper token naming)

> Note: Attempted to replace existing deployment scripts with a deployment script based off SuperfluidFrameworkDeployer, but this lead to far more headaches due to issues w/ library linking. This works out of the box with foundry, but not in JS environments.

### CODE REVIEW

**Comment by 0xdavinchee** [comment]
Taking some notes on subgraph integration issues w/ new script:
- the deployed contract addresses were previously and will continue to be hardcoded in: `ganache.json` and `addresses.template.ts`
- we can assume that deployment always starts from 0 nonce for CI so this should not be flakey (as long as no changes occur to the deploy scripts), but it would be nice to not have to restart everything after each run for local testing
This PR has already taken far too long so I am going to reference this issue https://github.com/superfluid-finance/protocol-monorepo/issues/1097 here so it doesn't get lost.

File: .github/workflows/call.setup-deploy-and-test-local-subgraph.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ on:
 
 jobs:
   setup-and-test-subgraph:
-    name: Build and test subgraph
+    name: Build and Test subgraph
```
Comment (hellwolf) [inline .github/workflows/call.setup-deploy-and-test-local-subgraph.yml:15]: o.O

File: packages/ethereum-contracts/contracts/utils/TestResolver.sol [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: AGPLv3
+pragma solidity 0.8.16;
+
+import { Resolver } from "./Resolver.sol";
+
+/// @title TestResolver contract
+/// @author Superfluid
+/// @notice A Resolver specifically used for testing
+/// @dev Used by the SuperfluidFrameworkDeployer to grant admin privileges to its deployer
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/utils/TestResolver.sol:9]: huh, why this is necessary?

File: .github/workflows/call.setup-deploy-and-test-local-subgraph.yml (hunk 1 above)
Comment (0xdavinchee) [reply .github/workflows/call.setup-deploy-and-test-local-subgraph.yml:15]: OCD 😅

File: packages/ethereum-contracts/package.json [hunk 3]
```diff
@@ -29,7 +29,7 @@
         "run-forge": "IS_FOUNDRY=true forge",
         "run-nodemon": "nodemon -e sol,js,ts -x",
         "build": "run-s build:*",
-        "build:contracts": "rm -f build/contracts/*;yarn run-truffle compile",
+        "build:contracts": "rm -f build/contracts/*;yarn run-truffle compile && yarn build:hardhat",
```
Comment (hellwolf) [inline packages/ethereum-contracts/package.json]: we are making buiild time longer?!

could split them into two:


build:contracts-truffle
build:contracts-hardhat

File: packages/ethereum-contracts/contracts/utils/TestResolver.sol (hunk 2 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/utils/TestResolver.sol:9]: `/// @dev Used by the SuperfluidFrameworkDeployer to grant admin privileges to its deployer`

File: packages/ethereum-contracts/package.json (hunk 3 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/package.json]: okay, true

File: packages/ethereum-contracts/scripts/deploy-test-framework.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
...5760006107ba565b815b925050506106e8565b600160a060020a0390811660009081526020818152604080832086845290915290205416905092915050565b6000808061081d857f01ffc9a70000000000000000000000000000000000000000000000000000000061094c565b909250905081158061082d575080155b1561083d576000925050506106e8565b61084f85600160e060020a031961094c565b909250905081158061086057508015155b15610870576000925050506106e8565b61087a858561094c565b909250905060018214801561088f5750806001145b1561089f576001925050506106e8565b506000949350505050565b600160a060020a0382166000908152600260209081526040808320600160e060020a03198516845290915281205460ff1615156108f2576108eb83836107ef565b90506106e8565b50600160a060020a03808316600081815260208181526040808320600160e060020a0319871684529091529020549091161492915050565b7bffffffffffffffffffffffffffffffffffffffffffffffffffffffff161590565b6040517f01ffc9a7000000000000000000000000000000000000000000000000000000008082526004820183905260009182919060208160248189617530fa90519096909550935050505056fea165627a7a72305820377f4a2d4301ede9949f163f319021a6e9c687c292a5e2b2c4734c126b524e6c00291ba01820182018201820182018201820182018201820182018201820182018201820a01820182018201820182018201820182018201820182018201820182018201820";
```
Comment (hellwolf) [inline packages/ethereum-contracts/scripts/deploy-test-framework.ts]: please reuse: ethereum-contracts/scripts/artifacts/ERC1820Registry.json

load json and prepend "0x" e.g.

File: packages/ethereum-contracts/scripts/deploy-test-token.js [hunk 5]
```diff
@@ -66,7 +66,8 @@ module.exports = eval(`(${S.toString()})()`)(async function (
         const testToken = await web3tx(TestToken.new, "TestToken.new")(
             tokenSymbol + " Fake Token",
             tokenSymbol,
-            tokenDecimals
+            tokenDecimals,
+            "1000000000000000000000000000000" // ethers.utils.parseUnits("1000000000000") 1e12
```
Comment (hellwolf) [inline packages/ethereum-contracts/scripts/deploy-test-token.js]: could use "1" + "0".repeat(30) if you want to be less errorprone here.

File: packages/ethereum-contracts/test/contracts/superfluid/SuperToken.NonStandard.test.ts [hunk 6]
```diff
@@ -218,6 +218,7 @@ describe("SuperToken's Non Standard Functions", function () {
                 "Test Token 6 Decimals",
                 "TEST6D",
                 6,
+                ethers.utils.parseUnits("1000000000000"),
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/superfluid/SuperToken.NonStandard.test.ts]: ditto.

File: packages/ethereum-contracts/test/contracts/superfluid/SuperTokenFactory.test.ts [hunk 7]
```diff
@@ -38,7 +38,8 @@ describe("SuperTokenFactory Contract", function () {
         token1 = await web3tx(TestToken.new, "TestToken.new 1")(
             "Test Token 1",
             "TT1",
-            18
+            18,
+            ethers.utils.parseUnits("1000000000000")
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/superfluid/SuperTokenFactory.test.ts]: ditto.

File: .github/workflows/call.setup-deploy-and-test-local-subgraph.yml (hunk 1 above)
Comment (hellwolf) [reply .github/workflows/call.setup-deploy-and-test-local-subgraph.yml:15]: image

File: packages/ethereum-contracts/contracts/utils/TestResolver.sol (hunk 2 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/utils/TestResolver.sol:9]: when you deploy Resolver from SuperfluidFrameworkDeployer, SuperfluidFrameworkDeployer is the admin, not the deployer of SuperfluidFrameworkDeployer, this allows the deployer of SuperfluidFrameworkDeployer to set things in the resolver at will without having to send funds to the SFDeployer contract, prank/impersonate as the SuperfluidFrameworkDeployer contract and then operate on it.

Comment (hellwolf) [reply packages/ethereum-contracts/contracts/utils/TestResolver.sol:9]: But deployer could set additional permissions, would that help?

Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/utils/TestResolver.sol:9]: yeah, we could always allow users to execute setting things on the resolver via the deployer contract if that's what you mean

File: .github/workflows/call.setup-deploy-and-test-local-subgraph.yml [hunk 8]
```diff
@@ -31,6 +31,12 @@ jobs:
       - name: "Install dependencies"
         run: yarn install --frozen-lockfile
 
+      - name: "Build contracts"
+        run: |
+          yarn build:truffle
```
Comment (hellwolf) [inline .github/workflows/call.setup-deploy-and-test-local-subgraph.yml]: build:contracts-truffle
build:contracts-hardhat

for reading clarity please.

**Comment by 0xdavinchee** [comment]
this is super flakey, going to work on the vestooor stuff for now

File: packages/ethereum-contracts/package.json [hunk 9]
```diff
@@ -29,8 +29,9 @@
         "run-forge": "IS_FOUNDRY=true forge",
         "run-nodemon": "nodemon -e sol,js,ts -x",
         "build": "run-s build:*",
-        "build:contracts": "rm -f build/contracts/*;yarn run-truffle compile",
-        "build:hardhat": "yarn run-hardhat compile",
+        "build:contracts": "rm -f build/contracts/*;yarn run-truffle compile && yarn build:contracts-hardhat",
```
Comment (hellwolf) [inline packages/ethereum-contracts/package.json]: `yarn run-truffle compile` -> `yarn build:contracts-truffle`

Comment (hellwolf) [reply packages/ethereum-contracts/package.json]: what's the output files of `yarn build:contracts-hardhat` ?

Comment (0xdavinchee) [reply packages/ethereum-contracts/package.json]: /artifacts similar to /build from truffle, but the format is _slightly_ different

**Review by hellwolf — APPROVED** [review]

**Pull request merged**